### PR TITLE
build: fix building libconsensus with reduced exports for Darwin targets

### PR DESCRIPTION
**Pull request by fanquake**
Darwin targets do not have a `protected` visibility function attribute, see [LLVM explanation](https://github.com/llvm/llvm-project/blob/8e9a505139fbef7d2e6e9d0adfe1efc87326f9ef/clang/lib/Basic/Targets/OSTargets.h#L131). This means that the `AX_GCC_FUNC_ATTRIBUTE` check for `visibility` fails:
```bash
configure:24513: checking for __attribute__((visibility))
configure:24537: g++ -std=c++11 -o conftest -g -O2  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS -DMAC_OSX -DOBJC_OLD_DISPATCH_PROTOTYPES=0  -Wl,-headerpad_max_install_names conftest.cpp  >&5
conftest.cpp:35:56: warning: target does not support 'protected' visibility; using 'default' [-Wunsupported-visibility]
                    int foo_pro( void ) __attribute__((visibility("protected")));
                                                       ^
1 warning generated.
configure:24537: $? = 0
configure:24550: result: no
```

This leads to `EXPORT_SYMBOL` being [defined to nothing](https://github.com/bitcoin/bitcoin/blob/f4de89edfa8be4501534fec0c662c650a4ce7ef2/src/script/bitcoinconsensus.h#L29), as `HAVE_FUNC_ATTRIBUTE_VISIBILITY` is not defined, and when building with reduced exports, you end up with a libbitcoinconsensus.dylib that doesn't export any  `_bitcoinconsensus_*` symbols.
```bash
➜  git:(master) nm -C src/.libs/libbitcoinconsensus.dylib | rg _bitcoinconsensus_
➜  git:(master)
```

We do have a [second check](https://github.com/bitcoin/bitcoin/blob/f4de89edfa8be4501534fec0c662c650a4ce7ef2/configure.ac#L882) for the `visibility` attribute, which works for Darwin as it's only testing for default visibility, however the result of this check isn't used at all. It was added in #4725, along with the `--enable-reduce-exports` option, however when libbitcoinconsensus was added in #5235, it used the results of the added `AX_GCC_FUNC_ATTRIBUTE` calls.

This PR removes our usage of the AX_GCC_FUNC_ATTRIBUTE macro entirely, in favour of our own checks in configure. This meant adding a check for `dllexport`, which I've tested as working with both [GCC](https://gcc.gnu.org/onlinedocs/gcc/Microsoft-Windows-Function-Attributes.html) and [Clang](https://releases.llvm.org/10.0.0/tools/clang/docs/AttributeReference.html#dllexport) when building for Windows. I haven't added an equivalent check for `dllimport`, as we weren't actually using the result of that check, we're just testing that `MSC_VER` was defined before using.

With these changes building a libbitcoinconsensus with reduced exports, when targeting Darwin, works as expected:
```bash
./autogen.sh
./configure --disable-tests --disable-bench --with-utils=no --with-daemon=no --with-gui=no --disable-wallet --with-libs=yes --enable-reduce-exports
make -j8
...
nm -C src/.libs/libbitcoinconsensus.dylib | rg _bitcoinconsensus_
000000000000a340 T _bitcoinconsensus_verify_script
00000000000097e0 T _bitcoinconsensus_verify_script_with_amount
000000000000a3c0 T _bitcoinconsensus_version
```

```python
>>> import ctypes
>>> consensus = ctypes.CDLL("src/.libs/libbitcoinconsensus.dylib")
>>> print(consensus.bitcoinconsensus_version())
1
>>> exit()
```

TODO: Modify a CI job to compile with --enable-reduce-exports and check for symbols in shared lib?

### CODE REVIEW

**Comment by fanquake** [comment]
A PR is never truly open until a moment of realization right after hitting the big green button.

This is actually something I broke in #7711. Updating the `AX_GCC_FUNC_ATTRIBUTE` macro reintroduced the checks for `protected` and `internal`, which had been removed in ee64c53c1fdcadf00138b7e6fdf44a1c433db9f8. However I take it for the past 4 years, no-one has built and used a libbitcoinconsensus for a Darwin target, with `--enable-reduced-exports`.

The alternative approach here would be to re-remove the two visibility attributes we don't care about from the `AX_GCC_FUNC_ATTRIBUTE` macro (rather than 7a600331a84fa7da0759f89deb309e60e5040773, 1381e9c8d327a86cbb5d2405d24b7ddd5b26fcd5 and 01bc0b6ca2b7b2d2b384f667e2abe7f1a7b9c8a5), noting that we've modified it from upstream, and then remove the unused visibility check from #4725. In that case, the remaining changes would still be relevant improvements.

Big anti-props me.

**Comment by fanquake** [comment]
Consensus from the latest build meeting was that removing the macro would be ok, so I'm going to leave this with the current approach. I've also modified the last commit to re-split the `RE_CXXFLAGS`, and conditionally output them and the `RELDFLAGS` at the end of configure.

**Comment by theuni** [comment]
Concept ACK, changes look good at a glance.

Thanks for fixing this. The attribute check has been broken for years.

Any reason for not just appending to CXXFLAGS/LDFLAGS for the configure output? I'm not sure that RE_FOO would anything to anyone other than reviewers of this pull request :)

**Comment by fanquake** [comment]
> Any reason for not just appending to CXXFLAGS/LDFLAGS for the configure output? I'm not sure that RE_FOO would anything to anyone other than reviewers of this pull request :)

Yea fair call. I've dropped the RE specifics and have added to the CXX and LD flags output.

File: src/script/bitcoinconsensus.h [hunk 1]
```diff
@@ -12,13 +12,13 @@
 #include <config/bitcoin-config.h>
   #if defined(_WIN32)
     #if defined(DLL_EXPORT)
-      #if defined(HAVE_FUNC_ATTRIBUTE_DLLEXPORT)
+      #if defined(HAVE_DLLEXPORT_ATTRIBUTE)
```
Comment (laanwj) [inline src/script/bitcoinconsensus.h:15]: We could also always assume `__declspec(dllexport)` to be present on windows? It seems something is very wrong with the compiler if it is not.

Comment (theuni) [reply src/script/bitcoinconsensus.h:15]: > We could also always assume __declspec(dllexport) to be present on windows? It seems something is very wrong with the compiler if it is not.

Agree. Also, this is our internal build, so we're generally aware of what we support, and we wouldn't support such a broken compiler.

Comment (theuni) [reply src/script/bitcoinconsensus.h:15]: In that case, the dllexport can be moved to a Windows-only check in configure, which errors if it fails.

**Comment by laanwj** [comment]
Code review ACK de4238f92f4c067f099663f68d9772105de81d75
nice diff +26 −246

**Pull request merged**